[11.0] stock_available_mrp: consider product types

ajaniszewska-dev opened the pull request:
Impacting module: stock_available_mrp 

potential_qty changed to -1.0 when BoM consist only from products type consumable or service (not stockable product type). Otherwise calculate potential_qty based on stockable product.

![quantity](https://user-images.githubusercontent.com/59824990/76979569-66ae3780-6938-11ea-89f7-6231c9af3555.png)

ajaniszewska-dev committed:
[11.0] stock_available_mrp: consider product types

gurneyalex commented on stock_available_mrp/models/product_product.py:
```diff
@@ -94,11 +94,11 @@ def _compute_available_quantities_dict(self):
             exploded_components = exploded_boms[product.id]
             component_needs = product._get_components_needs(
-                exploded_components)
+                exploded_components, True)
```
pass the name of the kwarg:
```suggestion
                exploded_components, only_stockable=True)
```

gurneyalex commented on stock_available_mrp/models/product_product.py:
```diff
@@ -144,6 +144,9 @@ def _get_components_needs(self, exploded_components):
         for bom_component in exploded_components:
             component = bom_component[0].product_id
-            needs += Counter({component: bom_component[1]['qty']})
+            if only_stockable and component.type != "product": #ignore 'service' and 'consu' in BoM to set potential_qty=-1.0
```
move the comment to line above the `if` statement, and check that the line length is < 80char

gurneyalex commented on stock_available_mrp/models/product_product.py:
```diff
@@ -144,6 +144,9 @@ def _get_components_needs(self, exploded_components):
-            needs += Counter({component: bom_component[1]['qty']})
+            if only_stockable and component.type != "product": #ignore 'service' and 'consu' in BoM to set potential_qty=-1.0
+                continue
+            else: 
```
you can remove the `else:` statement and unindent the next line by 1 level

gurneyalex reviewed (approved):
some improvements suggested, but i'm ok with the implementation. 

Please update the readme/DESCRIPTION.rst to indicate that -1 means "no limit to production"

If you can, add 2 tests:
* bom with 1 service product and 1 stockable product
* bom with 1 service product and 1 consumable product

ajaniszewska-dev commented:
> some improvements suggested, but i'm ok with the implementation.
> 
> Please update the readme/DESCRIPTION.rst to indicate that -1 means "no limit to production"
> 
> If you can, add 2 tests:
> 
>     * bom with 1 service product and 1 stockable product
> 
>     * bom with 1 service product and 1 consumable product

@gurneyalex Thank you for suggestion, updated README.rst with explanation regarding potential_qty = -1.0 and updated test file with 2 new test cases.

grindtildeath commented on stock_available_mrp/tests/test_potential_qty.py:
```diff
@@ -454,6 +454,42 @@ def test_potential_qty_list(self):
+        )
+
+    def test_potential_quantity_service_and_stockable(self):
+        p1 = self.product_model.create({'name': 'Test Product', 'type': 'product'})
```
Nitpicking: Use var names with meaning
```suggestion
        stockable_product = self.product_model.create({'name': 'Test Product', 'type': 'product'})
```

grindtildeath commented on stock_available_mrp/tests/test_potential_qty.py:
```diff
@@ -454,6 +454,42 @@ def test_potential_qty_list(self):
+
+    def test_potential_quantity_service_and_stockable(self):
+        p1 = self.product_model.create({'name': 'Test Product', 'type': 'product'})
+        p2 = self.product_model.create({'name': 'Test PService', 'type': 'service'})
```
Nitpicking: Use var names with meaning
```suggestion
        service_product = self.product_model.create({'name': 'Test PService', 'type': 'service'})
```

grindtildeath commented on stock_available_mrp/tests/test_potential_qty.py:
```diff
@@ -454,6 +454,42 @@ def test_potential_qty_list(self):
+        p2 = self.product_model.create({'name': 'Test PService', 'type': 'service'})
+
+        # P1 need one P2
+        self.create_simple_bom(p1, p2)
```
Nitpicking: Use var names with meaning
```suggestion
        self.create_simple_bom(stockable_product, service_product)
```

grindtildeath commented on stock_available_mrp/tests/test_potential_qty.py:
```diff
@@ -454,6 +454,42 @@ def test_potential_qty_list(self):
+        self.assertEqual(
+            {p1.id: -1.0, p2.id: 0.0},
+            {p.id: p.potential_qty for p in products}
+        )
```
NItpicking, but IMO you can simply do:
```suggestion
        self.assertEqual(stockable_product.potential_qty, -1.0)
```
No need to check the potential qty of a service product...

grindtildeath reviewed (approved):
Looks good, just some nitpicking

rousseldenis reviewed (approved):
Code review

simahawk commented:
/ocabot merge minor
